### PR TITLE
DWARF unit testing 

### PR DESCRIPTION
**Pull request by HoundThe**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**
#16846
Starting point for the dwarf tests, will be making further additions. The expectations are based of the `readelf` tool to get the dwarf information or 'dwarfdump'

### CODE REVIEW

**Comment by HoundThe** [comment]
I am thinking about the way to test it, I probably can't possibly hardcode check every attribute etc., I think probably cherry pick some abbreviations to check completely and just check the basics (number of attributes, has_children) etc at the others?

**Comment by HoundThe** [comment]
Afaik, we're only using line information as of now, but I also tested the abbreviations, we're gonna need it soon anyways. If this is a fine way to test, I'll do it aswell on a compiled C++ code, as for some I can't get line information parsed properly for C++, so I'll try to figure out that one and add it to the unit tests.

**Comment by HoundThe** [comment]
So I probably found a bug where the Line Number statements for DWARF aren't parsed for a C++ binary

> readelf -w cpp_dwarf3
```
....
 Line Number Statements:
  [0x00000029]  Set column to 3
  [0x0000002b]  Extended opcode 2: set Address to 0x11ee
  [0x00000036]  Special opcode 6: advance Address by 0 to 0x11ee and Line by 1 to 2
  [0x00000037]  Set column to 12
  [0x00000039]  Special opcode 173: advance Address by 12 to 0x11fa and Line by 0 to 2
  [0x0000003a]  Set column to 15
  [0x0000003c]  Special opcode 201: advance Address by 14 to 0x1208 and Line by 0 to 2
  [0x0000003d]  Advance PC by 3 to 0x120b
  [0x0000003f]  Extended opcode 1: End of Sequence

  [0x00000042]  Set column to 11
  [0x00000044]  Extended opcode 2: set Address to 0x120c
  [0x0000004f]  Special opcode 7: advance Address by 0 to 0x120c and Line by 2 to 3
  [0x00000050]  Set column to 21
  [0x00000052]  Special opcode 173: advance Address by 12 to 0x1218 and Line by 0 to 3
  [0x00000053]  Set column to 22
  [0x00000055]  Special opcode 201: advance Address by 14 to 0x1226 and Line by 0 to 3
  [0x00000056]  Advance PC by 3 to 0x1229
  [0x00000058]  Extended opcode 1: End of Sequence

  [0x0000005b]  Set column to 11
....
```

> r2 cpp_dwarf - `id`
```
....
Set column to 3
Extended opcode 2: set Address to 0x11ee
Special opcode 6: advance Address by 0 to 11ee and Line by 1 to 2
Set column to 12
Special opcode 173: advance Address by 12 to 11fa and Line by 0 to 2
Set column to 15
Special opcode 201: advance Address by 14 to 1208 and Line by 0 to 2
Advance PC by 3 to 0x120b
Extended opcode 1: End of Sequence
0x00001208	.//main.cpp	2
....
```

It looks like it stopts at the first sequence and doesn't continue, should I fix this before I move on with more tests?

**Comment by HoundThe** [comment]
Made issue about it #16975 16975

File: test/unit/test_dwarf.c [hunk 1]
```diff
@@ -0,0 +1,720 @@
...
+	r_io_bind (io, &bin->iob);
+
+	RBinOptions opt = { 0 };
+	bool res = r_bin_open (bin, "/home/hound/r2test/dwarf/c/dwarf3", &opt);
+	mu_assert ("couldn't open dwarf3", res);
+
+	// this is probably ugly, but I didn't know how to
+	// tell core  what bin to open so I did it myself
+	r_bin_free (core->bin);
+	core->bin = bin;
```
Comment (ret2libc) [inline test/unit/test_dwarf.c:42]: Why do you need core involved at all? Can't you just use `RBin` + `RIO`?

Comment (HoundThe) [reply test/unit/test_dwarf.c:42]: Yes, you're right. When I started this testing I was started with looking at bin_dwarf that takes RCore as an argument and didn't look back.

**Comment by HoundThe** [comment]
I've added, tests on a very basic binary, so a lot of times tests doesn't even change across the versions. I don't know the formats well enough to think of edge cases etc. right now, I might find some examples to test on. Also malformed files would be cool to do some tests. There is test on DWARF 5 which we don't really even parse yet. But I am counting on that these tests will be extend as I am going to continue to work on the parsing and getting to know the format better.

**Comment by HoundThe** [comment]
I was thinking about mergin the binaries https://github.com/radareorg/radare2-testbins/pull/23 and then merging this, so I can finish this PR and then open separate one for unit tests for debug_info and dwarf4, 5 additions

**Comment by XVilka** [comment]
I merged bins.

**Comment by HoundThe** [comment]
Oh I forgot, for DWARF 4 test to work, this needs to be merged aswell https://github.com/radareorg/radare2/pull/17061

**Comment by XVilka** [comment]
Please rebase.

File: test/unit/test_dwarf.c [hunk 2]
```diff
@@ -0,0 +1,997 @@
...
+	mu_assert_eq (da->decls[0].tag, DW_TAG_compile_unit, "Incorrect abbreviation");
+	{
+		mu_assert_eq (da->decls[0].has_children, 1, "Incorrect children flag");
+		// specs length is 8, because we don't parse
+		// DW_AT value: 0     DW_FORM value: 0
+		// so we have just 7 attrs below
+		mu_assert_eq (da->decls[0].length, 8, "Incorrect number of attributes");
+		{
+			int i = 0;
+			mu_assert_eq (da->decls[0].specs[i].attr_name, DW_AT_producer, "Incorrect children flag");
```
Comment (XVilka) [inline test/unit/test_dwarf.c:56]: Can be compactified with a macro, so only `DW_AT_*`/`DW_FORM_*` would change.

File: test/unit/test_dwarf.c [hunk 3]
```diff
@@ -0,0 +1,997 @@
...
+			mu_assert_eq (da->decls[0].specs[i].attr_form, DW_FORM_addr, "Incorrect children flag");
+			i++;
+			mu_assert_eq (da->decls[0].specs[i].attr_name, DW_AT_high_pc, "Incorrect children flag");
+			mu_assert_eq (da->decls[0].specs[i].attr_form, DW_FORM_addr, "Incorrect children flag");
+			i++;
+			mu_assert_eq (da->decls[0].specs[i].attr_name, DW_AT_stmt_list, "Incorrect children flag");
+			mu_assert_eq (da->decls[0].specs[i].attr_form, DW_FORM_data4, "Incorrect children flag");
+		}
+	}
+	mu_assert_eq (da->decls[1].tag, DW_TAG_variable, "Incorrect abbreviation");
```
Comment (XVilka) [inline test/unit/test_dwarf.c:78]: These can be compactified too.
